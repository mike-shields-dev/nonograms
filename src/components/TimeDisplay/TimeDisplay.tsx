interface Props {
  durationMs?: number;
  showMilliseconds?: boolean;
}

export default function TimeDisplay({
  durationMs,
  showMilliseconds = false,
}: Props) {
  const [hours, minutes, seconds, milliseconds] = [
    {
      name: "hours",
      value: 0,
      pad: {
        length: 2,
        char: "0",
      },
      units: "h",
    },
    {
      name: "minutes",
      value: 0,
      pad: {
        length: 2,
        char: "0",
      },
      units: "m",
    },
    {
      name: "seconds",
      value: 0,
      pad: {
        length: 2,
        char: "0",
      },
      units: "s",
    },
    {
      name: "milliseconds",
      value: 0,
      pad: {
        length: 3,
        char: "0",
      },
      units: "ms",
    },
  ];

  if (typeof durationMs === "number") {
    hours.value = Math.floor((durationMs / 1000 / 3600) % 24);
    minutes.value = Math.floor((durationMs / 1000 / 60) % 60);
    seconds.value = Math.floor((durationMs / 1000) % 60);
    milliseconds.value = showMilliseconds ? durationMs % 1000 : 0;
  }

  return (
    <span>
      {(showMilliseconds
        ? [hours, minutes, seconds, milliseconds]
        : [hours, minutes, seconds]
      ).map(({ value, name, units, pad: { length, char } }, i, array) => (
        <span key={`time-display-${name}`}>
          <span data-testid={`time-display-${name}`}>
            {`${value}`.padStart(length, char)}
          </span>
          <span>{units}</span>
          {i < array.length - 1 && <span>:</span>}
        </span>
      ))}
    </span>
  );
}
