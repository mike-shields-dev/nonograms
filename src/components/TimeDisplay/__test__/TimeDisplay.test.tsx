import { describe, it, expect, afterEach } from "vitest";
import { cleanup, render, screen } from "@testing-library/react";
import TimeDisplay from "../TimeDisplay";

describe("TimeDisplay", () => {
  afterEach(() => cleanup());

  describe("with showMilliseconds prop equal to true", () => {
    it("matches snapshot", () => {
      const { container } = render(
        <TimeDisplay durationMs={0} showMilliseconds={true} />
      );

      expect(container).toMatchSnapshot();
    });

    hoursTestCases.forEach(({ durationMs, expected }) => {
      it("renders hours correctly", () => {
        render(<TimeDisplay durationMs={durationMs} showMilliseconds={true} />);

        expect(screen.getByTestId("time-display-hours").textContent).toBe(
          expected.hours
        );
      });
    });

    minutesTestCases.forEach(({ durationMs, expected }) => {
      it("renders minutes correctly", () => {
        render(<TimeDisplay durationMs={durationMs} showMilliseconds={true} />);

        expect(screen.getByTestId("time-display-minutes").textContent).toBe(
          expected.minutes
        );
      });
    });

    secondsTestCases.forEach(({ durationMs, expected }) => {
      it("renders seconds correctly", () => {
        render(<TimeDisplay durationMs={durationMs} showMilliseconds={true} />);

        expect(screen.getByTestId("time-display-seconds").textContent).toBe(
          expected.seconds
        );
      });

      cleanup();
    });

    millisecondsTestCases.forEach(({ durationMs, expected }) => {
      it("renders milliseconds correctly", () => {
        render(<TimeDisplay durationMs={durationMs} showMilliseconds={true} />);

        expect(
          screen.getByTestId("time-display-milliseconds").textContent
        ).toBe(expected.milliseconds);
      });
    });
  });

  describe("with showMilliseconds prop equal to false", () => {
    it("matches snapshot", () => {
      const { container } = render(
        <TimeDisplay durationMs={0} showMilliseconds={false} />
      );

      expect(container).toMatchSnapshot();
    });

    it("does not display milliseconds", () => {
      render(<TimeDisplay durationMs={0} showMilliseconds={false} />);

      expect(screen.queryByTestId("time-display-milliseconds")).toBeNull();
    });

    hoursTestCases.forEach(({ durationMs, expected }) => {
      it("renders hours correctly", () => {
        render(
          <TimeDisplay durationMs={durationMs} showMilliseconds={false} />
        );

        expect(screen.getByTestId("time-display-hours").textContent).toBe(
          expected.hours
        );
      });
    });

    minutesTestCases.forEach(({ durationMs, expected }) => {
      it("renders minutes correctly", () => {
        render(
          <TimeDisplay durationMs={durationMs} showMilliseconds={false} />
        );

        expect(screen.getByTestId("time-display-minutes").textContent).toBe(
          expected.minutes
        );
      });
    });

    secondsTestCases.forEach(({ durationMs, expected }) => {
      it("renders seconds correctly", () => {
        render(
          <TimeDisplay durationMs={durationMs} showMilliseconds={false} />
        );

        expect(screen.getByTestId("time-display-seconds").textContent).toBe(
          expected.seconds
        );
      });
    });
  });
});

const millisecondsTestCases = [
  {
    durationMs: 0,
    expected: {
      milliseconds: "000",
      seconds: "00",
      minutes: "00",
      hours: "00",
    },
  },
  {
    durationMs: 1,
    expected: {
      milliseconds: "001",
      seconds: "00",
      minutes: "00",
      hours: "00",
    },
  },
  {
    durationMs: 9,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "00",
      milliseconds: "009",
    },
  },
  {
    durationMs: 10,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "00",
      milliseconds: "010",
    },
  },
  {
    durationMs: 99,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "00",
      milliseconds: "099",
    },
  },
  {
    durationMs: 100,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "00",
      milliseconds: "100",
    },
  },
  {
    durationMs: 999,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "00",
      milliseconds: "999",
    },
  },
];

const secondsTestCases = [
  {
    durationMs: 1000,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "01",
      milliseconds: "000",
    },
  },
  {
    durationMs: 9000,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "09",
      milliseconds: "000",
    },
  },
  {
    durationMs: 10000,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "10",
      milliseconds: "000",
    },
  },
  {
    durationMs: 59000,
    expected: {
      hours: "00",
      minutes: "00",
      seconds: "59",
      milliseconds: "000",
    },
  },
];

const minutesTestCases = [
  {
    durationMs: 60000,
    expected: {
      hours: "00",
      minutes: "01",
      seconds: "00",
      milliseconds: "000ms",
    },
  },
  {
    durationMs: 3540000,
    expected: {
      hours: "00",
      minutes: "59",
      seconds: "00",
      milliseconds: "000",
    },
  },
];

const hoursTestCases = [
  {
    durationMs: 3600000,
    expected: {
      hours: "01",
      minutes: "00",
      seconds: "00",
      milliseconds: "000",
    },
  },
  {
    durationMs: 82800000,
    expected: {
      hours: "23",
      minutes: "00",
      seconds: "00",
      milliseconds: "000",
    },
  },
];
