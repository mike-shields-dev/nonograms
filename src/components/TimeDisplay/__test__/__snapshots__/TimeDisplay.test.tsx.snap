// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeDisplay > with showMilliseconds prop equal to false > matches snapshot 1`] = `
<div>
  <span>
    <span>
      <span
        data-testid="time-display-hours"
      >
        00
      </span>
      <span>
        h
      </span>
      <span>
        :
      </span>
    </span>
    <span>
      <span
        data-testid="time-display-minutes"
      >
        00
      </span>
      <span>
        m
      </span>
      <span>
        :
      </span>
    </span>
    <span>
      <span
        data-testid="time-display-seconds"
      >
        00
      </span>
      <span>
        s
      </span>
    </span>
  </span>
</div>
`;

exports[`TimeDisplay > with showMilliseconds prop equal to true > matches snapshot 1`] = `
<div>
  <span>
    <span>
      <span
        data-testid="time-display-hours"
      >
        00
      </span>
      <span>
        h
      </span>
      <span>
        :
      </span>
    </span>
    <span>
      <span
        data-testid="time-display-minutes"
      >
        00
      </span>
      <span>
        m
      </span>
      <span>
        :
      </span>
    </span>
    <span>
      <span
        data-testid="time-display-seconds"
      >
        00
      </span>
      <span>
        s
      </span>
      <span>
        :
      </span>
    </span>
    <span>
      <span
        data-testid="time-display-milliseconds"
      >
        000
      </span>
      <span>
        ms
      </span>
    </span>
  </span>
</div>
`;
